src/06_swap_chain_creation.rs

```diff
@@ -0,0 +1,185 @@
+#[cfg(feature = "dx12")]
+extern crate gfx_backend_dx12 as back;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "vulkan")]
+extern crate gfx_backend_vulkan as back;
+extern crate gfx_hal as hal;
+extern crate winit;
+
+use hal::{
+    Instance,
+    Device,
+    QueueFamily,
+    PhysicalDevice,
+    Capability,
+    Surface,
+    SwapchainConfig,
+};
+
+static WINDOW_NAME: &str = "06_swap_chain_creation";
+
+fn main() {
+    let (window, events_loop) = init_window();
+    let instance = init_hal();
+    let mut surface = create_surface(&instance, &window);
+    let mut adapter = pick_adapter(&instance);
+    let (device, _command_queues) =
+        create_device_with_graphics_queues(&mut adapter);
+    let (_swapchain, _backbuffer, _extent, _format) =
+        create_swap_chain(&adapter, &device, &mut surface, &window, None);
+    main_loop(events_loop);
+}
+
+fn create_swap_chain(adapter: &hal::Adapter<back::Backend>,
+                     device: &<back::Backend as hal::Backend>::Device,
+                     surface: &mut <back::Backend as hal::Backend>::Surface,
+                     window: &winit::Window,
+                     previous_swapchain: Option<<back::Backend as hal::Backend>::Swapchain>)
+    -> (<back::Backend as hal::Backend>::Swapchain,
+        hal::Backbuffer<back::Backend>,
+        hal::window::Extent2D, hal::format::Format) {
+
+    let (caps, formats, _present_modes) = surface.compatibility(&adapter.physical_device);
+
+    let format = formats
+        .map_or(hal::format::Format::Rgba8Srgb, |formats|
+            formats
+                .iter()
+                .find(|format| format.base_format().1 == hal::format::ChannelType::Srgb)
+                .map(|format| *format)
+                .unwrap_or(formats[0])
+        );
+
+    let extent = match caps.current_extent {
```

Comment:
Maybe could be simplified a bit, using something like:
```rust
let extent = caps.current_extent.unwrap_or_else(|| {
    let window_size: (u32, u32) = window
        .get_inner_size()
        .unwrap()
        .to_physical(window.get_hidpi_factor())
        .into();

    let std::ops::Range { start, end } = caps.extents;

    hal::window::Extent2D {
        width: window_size.0.max(start.width).min(end.width),
        height: window_size.1.max(start.height).min(end.height),
    }
});
```

Reply:
neat!